File: framework/languages/java/e2e/ResilienceLifecycle/Server/Consumer/src/main/java/systems/zlink/e2e/resiliencelifecycle/consumer/endpoints/ConsumerEndpoints.java
package systems.zlink.e2e.resiliencelifecycle.consumer.endpoints;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.concurrent.Executors;
import org.springframework.context.SmartLifecycle;
import systems.zlink.e2e.resiliencelifecycle.consumer.scenarios.ConsumerScenario;

public final class ConsumerEndpoints implements SmartLifecycle {
    private final ObjectMapper json;
    private final ConsumerScenario scenario;
    private final String endpoint;
    private HttpServer server;
    private boolean running;

    public ConsumerEndpoints(ObjectMapper json, ConsumerScenario scenario, String endpoint) {
        this.json = json;
        this.scenario = scenario;
        this.endpoint = endpoint;
    }

    @Override
    public void start() {
        try {
            URI uri = URI.create(endpoint);
            server = HttpServer.create(new InetSocketAddress(uri.getHost(), uri.getPort()), 0);
            server.setExecutor(Executors.newCachedThreadPool());
            server.createContext("/health", exchange -> writeJson(exchange, Map.of("status", "ready")));
            server.createContext("/scenario", exchange -> {
                String mode = exchange.getRequestURI().getPath().replaceFirst("^/scenario/?", "");
                if (mode.isBlank()) {
                    mode = "default";
                }
                scenario.run(mode);
                writeJson(exchange, Map.of("status", "passed", "mode", mode));
            });
            server.start();
            running = true;
        } catch (Exception error) {
            throw new IllegalStateException("failed to start consumer endpoint " + endpoint, error);
        }
    }

    @Override
    public void stop() {
        if (server != null) {
            server.stop(0);
            server = null;
        }
        running = false;
    }

    @Override
    public boolean isRunning() {
        return running;
    }

    private void writeJson(com.sun.net.httpserver.HttpExchange exchange, Object value) throws java.io.IOException {
        try {
            byte[] body = json.writeValueAsBytes(value);
            exchange.getResponseHeaders().add("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, body.length);
            exchange.getResponseBody().write(body);
        } catch (RuntimeException error) {
            byte[] body = error.toString().getBytes(StandardCharsets.UTF_8);
            exchange.sendResponseHeaders(500, body.length);
            exchange.getResponseBody().write(body);
        } finally {
            exchange.close();
        }
    }
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Server/Consumer/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/consumer/Endpoints/ConsumerEndpoints.kt
package systems.zlink.e2e.kotlin.registrymessaging.consumer.Endpoints

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.net.URI
import java.time.Duration
import java.util.concurrent.Executors
import org.springframework.context.ConfigurableApplicationContext
import systems.zlink.e2e.kotlin.registrymessaging.consumer.Configuration.ConsumerOptions
import systems.zlink.e2e.kotlin.registrymessaging.shared.Contracts
import systems.zlink.e2e.kotlin.registrymessaging.shared.PayloadRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.PayloadReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileMsg
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.RequestFailureRes
import systems.zlink.framework.channels.ZLinkClient

class ConsumerEndpoints(
    private val options: ConsumerOptions,
    private val context: ConfigurableApplicationContext,
) {
    private val mapper = jacksonObjectMapper()
    private val channels = context.getBean(ZLinkClient::class.java)

    fun start(): HttpServer {
        val uri = URI.create(options.httpUrl)
        val server = HttpServer.create(InetSocketAddress(uri.host, uri.port), 0)
        server.executor = Executors.newCachedThreadPool()
        server.createContext("/health") { exchange ->
            exchange.writeJson(mapOf("status" to "ready", "role" to "consumer"))
        }
        server.createContext("/profile/batch-request") { exchange ->
            val requests = exchange.readJson<Array<ProfileReq>>()
            exchange.writeJson(requests.map { requestProfile(it, Duration.ofSeconds(5)) })
        }
        server.createContext("/profile/request") { exchange ->
            exchange.writeJson(requestProfile(exchange.readJson(), Duration.ofSeconds(5)))
        }
        server.createContext("/profile/slow-request") { exchange ->
            val request = exchange.readJson<ProfileReq>()
            try {
                requestProfile(request, Duration.ofMillis(100), retryUntilReady = false)
                exchange.writeJson(RequestFailureRes(false, ""))
            } catch (error: RuntimeException) {
                exchange.writeJson(RequestFailureRes(true, rootName(error)))
            }
        }
        server.createContext("/profile/missing-request") { exchange ->
            val request = exchange.readJson<ProfileReq>()
            try {
                channels.requestToChannel(Contracts.PROFILE_CHANNEL, request)
                    .packetName("MissingProfileReq")
                    .timeout(Duration.ofSeconds(5))
                    .await(ProfileRes::class.java)
                exchange.writeJson(RequestFailureRes(false, ""))
            } catch (error: RuntimeException) {
                exchange.writeJson(RequestFailureRes(true, rootName(error)))
            }
        }
        server.createContext("/profile/missing-command") { exchange ->
            val command = exchange.readJson<ProfileMsg>()
            channels.sendToChannel(Contracts.PROFILE_CHANNEL, command)
                .packetName("MissingProfileMsg")
                .await()
            exchange.writeJson(mapOf("status" to "sent"))
        }
        server.createContext("/profile/payload") { exchange ->
            exchange.writeJson(requestPayload(exchange.readJson()))
        }
        server.createContext("/shutdown") { exchange ->
            exchange.writeJson(mapOf("status" to "stopping"))
            Thread {
                server.stop(0)
                context.close()
            }.start()
        }
        server.start()
        return server
    }

    private fun requestProfile(
        request: ProfileReq,
        timeout: Duration,
        retryUntilReady: Boolean = true,
    ): ProfileRes {
        val deadline = System.nanoTime() + Duration.ofSeconds(30).toNanos()
        var last: RuntimeException? = null
        while (System.nanoTime() < deadline) {
            try {
                return channels.requestToChannel(Contracts.PROFILE_CHANNEL, request)
                    .packetName(Contracts.PROFILE_REQUEST_PACKET)
                    .timeout(timeout)
                    .await(ProfileRes::class.java)
            } catch (error: RuntimeException) {
                if (!retryUntilReady) {
                    throw error
                }
                last = error
                Thread.sleep(100)
            }
        }
        throw IllegalStateException("Timed out waiting for profile endpoint.", last)
    }

    private fun requestPayload(request: PayloadReq): PayloadRes {
        val deadline = System.nanoTime() + Duration.ofSeconds(30).toNanos()
        var last: RuntimeException? = null
        while (System.nanoTime() < deadline) {
            try {
                return channels.requestToChannel(Contracts.PROFILE_CHANNEL, request)
                    .packetName(Contracts.PAYLOAD_REQUEST_PACKET)
                    .timeout(Duration.ofSeconds(10))
                    .await(PayloadRes::class.java)
            } catch (error: RuntimeException) {
                last = error
                Thread.sleep(100)
            }
        }
        throw IllegalStateException("Timed out waiting for payload endpoint.", last)
    }

    private inline fun <reified T> HttpExchange.readJson(): T =
        requestBody.use { mapper.readValue(it) }

    private fun HttpExchange.writeJson(value: Any) {
        try {
            val bytes = mapper.writeValueAsBytes(value)
            responseHeaders.add("content-type", "application/json")
            sendResponseHeaders(200, bytes.size.toLong())
            responseBody.use { it.write(bytes) }
        } catch (error: Exception) {
            val bytes = mapper.writeValueAsBytes(mapOf("error" to (error.message ?: error.javaClass.name)))
            sendResponseHeaders(500, bytes.size.toLong())
            responseBody.use { it.write(bytes) }
        }
    }

    private fun rootName(error: Throwable): String {
        var current = error
        while (current.cause != null) {
            current = current.cause!!
        }
        return current.javaClass.simpleName
    }
}
